core/vm/instructions.go

```diff
@@ -440,13 +442,19 @@ func opBlockhash(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) (
 	}
 	var upper, lower uint64
 	upper = interpreter.evm.Context.BlockNumber.Uint64()
-	if upper < 2 {
+	if upper < 257 {
 		lower = 0
 	} else {
-		lower = upper - 1
+		lower = upper - 256
 	}
 	if num64 >= lower && num64 < upper {
-		num.SetBytes(interpreter.evm.Context.GetHash(num64).Bytes())
+		chainId := interpreter.evm.ChainConfig().ChainID
+		hasher := sha3.NewLegacyKeccak256()
```

Comment:
Based on `opSha3` and https://github.com/ethereum/go-ethereum/pull/17863, it seems better to reuse `interpreter.hasher` here.